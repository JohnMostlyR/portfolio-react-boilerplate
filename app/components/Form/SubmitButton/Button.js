import styled from 'styled-components';
import PropTypes from 'prop-types';

import { iconUrl } from '../HandIcon';
import { theme } from '../theme';
import { BASE_FONT_BOLD } from '../typography';

const { button } = theme;

const Button = styled.button`
  position: relative;
  margin: 0 2rem;
  z-index: 2;
  appearance: none;
  background-color: transparent;
  border: 0;
  filter: ${({ isEnabled }) => (isEnabled ? 'none' : 'grayscale()')};
  padding: 0;
  ${BASE_FONT_BOLD} > span {
    display: inline-block;
    background-color: ${button.backgroundColor};
    border: 1px solid ${button.backgroundColor};
    border-radius: 5px;
    color: ${button.color};
    cursor: ${({ isEnabled }) => (isEnabled ? 'pointer' : 'not-allowed')};
    padding: 8px;
  }

  &::after,
  &::before {
    position: absolute;
    content: '';
    height: 32px;
    top: 50%;
    width: 32px;
    z-index: -1;
    background-image: ${() => `url(${iconUrl})`};
    background-position: center;
    background-repeat: no-repeat;
    background-size: 100%;
    color: inherit;
    transition: transform 0.3s ease-in-out;
  }

  &::after {
    transform: translate3d(-100%, -50%, 0);
  }

  &::before {
    transform: translate3d(0, -50%, 0) rotateY(180deg);
  }

  &:focus,
  &:hover {
    &::after {
      transform: translate3d(25%, -50%, 0);
    }

    &::before {
      transform: translate3d(-125%, -50%, 0) rotateY(180deg);
    }
  }
`;

Button.propTypes = {
  isEnabled: PropTypes.bool,
};

Button.defaultProps = {
  isEnabled: false,
};

export default Button;
