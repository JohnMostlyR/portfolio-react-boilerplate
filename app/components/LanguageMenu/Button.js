/**
 * Button Component
 * @param label
 * @param isExpanded
 * @param toggleMenuHandler
 * @param buttonRef
 * @returns {*}
 * @constructor
 */
import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import SVGIcon from '../SVGIcon';

const StyledButton = styled.button.attrs({
  lang: 'en',
  type: 'button',
})`
  &:hover {
    cursor: pointer;
  }

  @supports (display: flex) {
    align-items: center;
    display: flex;
  }
`;

const Label = styled.p`
  display: inherit;
  position: absolute;
  width: 1px;
  height: 1px;
  margin: -1px;
  padding: 0;
  border: 0;
  overflow: hidden;
  clip: rect(0 0 0 0);

  @media (min-width: 768px) {
    display: inline-block;
    height: auto;
    margin: 0 0 0 0.5em;
    overflow: visible;
    position: relative;
    width: auto;
    clip: initial;
    font-size: 1rem;
  }
`;

function Button({ label, isExpanded, toggleMenuHandler, buttonRef }) {
  function handleKeyDown(evt) {
    const { keyCode } = evt;

    switch (keyCode) {
      case 13: // Enter
      case 32: // Space
      case 40: // Arrow Down
        evt.preventDefault();
        toggleMenuHandler();
        break;
      default:
        break;
    }
  }

  function handleClick(evt) {
    evt.preventDefault();
    evt.stopPropagation();
    toggleMenuHandler();
  }

  return (
    <StyledButton
      onClick={handleClick}
      onKeyDown={handleKeyDown}
      aria-expanded={isExpanded}
      aria-haspopup="true"
      innerRef={buttonRef}
    >
      <SVGIcon viewBox="0 0 15.304 15.304">
        <g>
          <path d="M5.734 8.248v-.631a5.587 5.587 0 0 1-.992-.863 12.413 12.413 0 0 1-1.836 1.464l-.081.05-.49-.783.08-.051c.01-.007.876-.556 1.751-1.413a8.345 8.345 0 0 1-1.124-2.402l-.023-.09.895-.23.022.091c.002.01.257.954.877 1.926.613-.745.97-1.463 1.061-2.14H2.115V2.171h2.072v-.559h.966v.558h2.072v1.006h-.417c-.097.941-.575 1.916-1.424 2.901.114.128.231.245.35.357v-.473c0-.336.272-.608.608-.608h2.705V.569a.151.151 0 0 0-.152-.152H.152A.152.152 0 0 0 0 .569v8.742c0 .085.068.152.152.152h5.582V8.248zm4.994.267h-.017c-.084.333-.166.755-.258 1.078l-.332 1.187h1.236l-.347-1.187c-.1-.332-.2-.746-.282-1.078z" />
          <path d="M15.152 5.841H6.41a.152.152 0 0 0-.152.152v8.743c0 .084.068.151.152.151h8.742a.15.15 0 0 0 .152-.151V5.993a.152.152 0 0 0-.152-.152zm-3.18 7.319l-.431-1.436H9.94l-.399 1.436h-1.31L9.94 7.57h1.658l1.733 5.59h-1.359zM3.935 9.846H2.72v2.02l1.006.996h2.26v-1.215H3.935z" />
        </g>
      </SVGIcon>
      <Label>{label}</Label>
    </StyledButton>
  );
}

Button.propTypes = {
  label: PropTypes.string,
  isExpanded: PropTypes.bool,
  toggleMenuHandler: PropTypes.func.isRequired,
  buttonRef: PropTypes.func.isRequired,
};

Button.defaultProps = {
  label: 'Label',
  isExpanded: false,
};

export default Button;
