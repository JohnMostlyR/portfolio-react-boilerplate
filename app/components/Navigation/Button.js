import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { theme } from '../../styles/theme';
import { BASE_FONT_REGULAR, BASE_LINE_HEIGHT } from '../../styles/typography';

const { backgroundColor, color } = theme.site.navigation.link;

/* eslint-disable indent */
const StyledButton = styled.button.attrs({
  type: 'button',
})`
  display: inline-block;
  margin-left: auto;
  background-color: ${backgroundColor};
  border-radius: 0.25rem;
  box-shadow: 0 0 5px lightgrey;
  color: ${color};
  cursor: pointer;
  letter-spacing: normal;
  padding: 0.5rem;
  text-align: left;
  transition: box-shadow 0.3s 0s ease-in-out;
  white-space: nowrap;
  word-break: keep-all;
  line-height: ${BASE_LINE_HEIGHT};
  ${BASE_FONT_REGULAR} &:hover {
    box-shadow: 0 0 1px gray;
  }

  @media (min-width: ${({ bigScreenBreakpoint }) =>
      `${bigScreenBreakpoint}px`}) {
    display: none;
  }
`;

const SVGIcon = styled.svg.attrs({
  xmlns: 'http://www.w3.org/2000/svg',
  viewBox: '0 0 100 100',
  focusable: 'false',
})`
  height: 16px;
  position: relative;
  width: 16px;
  fill: currentColor;
`;

const SVGPath = styled.path`
  opacity: 1;
  transition: transform 0.3s ease-in-out, opacity 0.2s ease-in-out;
`;

/* stylelint-disable block-no-empty */
const SVGPathOne = styled(SVGPath).attrs({
  d: 'M5 13h141v14H5z',
  transform: ({ isExpanded }) =>
    isExpanded ? 'translate(16 -16) rotate(45 0 0)' : 'rotate(0) translate(0)',
})``;

const SVGPathTwo = styled(SVGPath).attrs({
  d: 'M5 43h90v14H5z',
})`
  opacity: ${({ isExpanded }) => (isExpanded ? 0 : 1)};
`;

const SVGPathThree = styled(SVGPath).attrs({
  d: 'M5 73h141v14H5z',
  transform: ({ isExpanded }) =>
    isExpanded
      ? 'translate(16, 16) rotate(-45 0 100)'
      : 'rotate(0) translate(0)',
})``;

const Label = styled.span`
  margin-left: 0.5em;
  vertical-align: middle;
`;

function Button({
  bigScreenBreakpoint,
  isExpanded,
  label,
  toggleMenu,
  buttonRef,
}) {
  function handleKeyDown(evt) {
    const { keyCode } = evt;

    switch (keyCode) {
      case 13: // Enter
      case 32: // Space
      case 40: // Arrow Down
        evt.stopPropagation();
        evt.preventDefault();
        toggleMenu();
        break;
      default:
        break;
    }
  }

  function handleClick(evt) {
    evt.stopPropagation();
    evt.preventDefault();
    toggleMenu();
  }

  return (
    <StyledButton
      aria-expanded={isExpanded}
      bigScreenBreakpoint={bigScreenBreakpoint}
      innerRef={buttonRef}
      onClick={handleClick}
      onKeyDown={handleKeyDown}
    >
      <SVGIcon>
        <SVGPathOne isExpanded={isExpanded} />
        <SVGPathTwo isExpanded={isExpanded} />
        <SVGPathThree isExpanded={isExpanded} />
      </SVGIcon>
      <Label>{label}</Label>
    </StyledButton>
  );
}

Button.propTypes = {
  bigScreenBreakpoint: PropTypes.number.isRequired,
  isExpanded: PropTypes.bool,
  label: PropTypes.string,
  toggleMenu: PropTypes.func.isRequired,
  buttonRef: PropTypes.func.isRequired,
};

Button.defaultProps = {
  label: 'Label',
  isExpanded: false,
};

export default Button;
