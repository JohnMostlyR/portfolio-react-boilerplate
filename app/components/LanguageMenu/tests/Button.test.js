import React from 'react';
import { shallow, mount } from 'enzyme';

import Button from '../Button';

describe('<Button />', () => {
  const toggleMenuHandler = jest.fn();
  const buttonRef = jest.fn();

  it('should render and match the snapshot', () => {
    const renderedComponent = shallow(
      <Button toggleMenuHandler={toggleMenuHandler} buttonRef={buttonRef} />,
    );

    expect(renderedComponent).toMatchSnapshot();
  });

  it('should adopt the "label" property', () => {
    const PROPERTY_VALUE = 'Test';
    const renderedComponent = shallow(
      <Button
        toggleMenuHandler={toggleMenuHandler}
        buttonRef={buttonRef}
        label={PROPERTY_VALUE}
      />,
    );

    expect(renderedComponent).toMatchSnapshot();
  });

  it('should adopt the "isExpanded" property', () => {
    const PROPERTY_VALUE = true;
    const renderedComponent = shallow(
      <Button
        toggleMenuHandler={toggleMenuHandler}
        buttonRef={buttonRef}
        isExpanded={PROPERTY_VALUE}
      />,
    );

    expect(renderedComponent.prop('aria-expanded')).toBe(PROPERTY_VALUE);
  });

  it('should adopt the "buttonRef" property', () => {
    mount(
      <Button toggleMenuHandler={toggleMenuHandler} buttonRef={buttonRef} />,
    );

    expect(buttonRef).toHaveBeenCalledTimes(1);
  });

  describe('The "toggleMenuHandler" property', () => {
    const PREVENT_DEFAULT = jest.fn();
    const STOP_PROPAGATION = jest.fn();
    const renderedComponent = shallow(
      <Button toggleMenuHandler={toggleMenuHandler} buttonRef={buttonRef} />,
    );

    afterEach(() => {
      jest.clearAllMocks();
    });

    it('should get called on a click event', () => {
      renderedComponent.simulate('click', {
        preventDefault: PREVENT_DEFAULT,
        stopPropagation: STOP_PROPAGATION,
      });
      expect(PREVENT_DEFAULT).toHaveBeenCalledTimes(1);
      expect(STOP_PROPAGATION).toHaveBeenCalledTimes(1);
      expect(toggleMenuHandler).toHaveBeenCalledTimes(1);
    });

    it('should get called when the "Enter" key is pressed', () => {
      renderedComponent.simulate('keyDown', {
        preventDefault: PREVENT_DEFAULT,
        keyCode: 13,
      });
      expect(PREVENT_DEFAULT).toHaveBeenCalledTimes(1);
      expect(toggleMenuHandler).toHaveBeenCalledTimes(1);
    });

    it('should get called when the "Space" key is pressed', () => {
      renderedComponent.simulate('keyDown', {
        preventDefault: PREVENT_DEFAULT,
        keyCode: 32,
      });
      expect(PREVENT_DEFAULT).toHaveBeenCalledTimes(1);
      expect(toggleMenuHandler).toHaveBeenCalledTimes(1);
    });

    it('should get called when the "Arrow Down" key is pressed', () => {
      renderedComponent.simulate('keyDown', {
        preventDefault: PREVENT_DEFAULT,
        keyCode: 40,
      });
      expect(PREVENT_DEFAULT).toHaveBeenCalledTimes(1);
      expect(toggleMenuHandler).toHaveBeenCalledTimes(1);
    });

    it('should NOT get called when any other key is pressed', () => {
      renderedComponent.simulate('keyDown', {
        preventDefault: PREVENT_DEFAULT,
        keyCode: 0,
      });
      expect(PREVENT_DEFAULT).not.toHaveBeenCalled();
      expect(toggleMenuHandler).not.toHaveBeenCalled();
    });
  });
});
