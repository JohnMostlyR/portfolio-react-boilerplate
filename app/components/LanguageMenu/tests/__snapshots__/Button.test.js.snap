// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should adopt the "label" property 1`] = `
<Button__StyledButton
  aria-expanded={false}
  aria-haspopup="true"
  innerRef={[MockFunction]}
  onClick={[Function]}
  onKeyDown={[Function]}
>
  <SVGIcon
    viewBox="0 0 15.304 15.304"
  >
    <g>
      <path
        d="M5.734 8.248v-.631a5.587 5.587 0 0 1-.992-.863 12.413 12.413 0 0 1-1.836 1.464l-.081.05-.49-.783.08-.051c.01-.007.876-.556 1.751-1.413a8.345 8.345 0 0 1-1.124-2.402l-.023-.09.895-.23.022.091c.002.01.257.954.877 1.926.613-.745.97-1.463 1.061-2.14H2.115V2.171h2.072v-.559h.966v.558h2.072v1.006h-.417c-.097.941-.575 1.916-1.424 2.901.114.128.231.245.35.357v-.473c0-.336.272-.608.608-.608h2.705V.569a.151.151 0 0 0-.152-.152H.152A.152.152 0 0 0 0 .569v8.742c0 .085.068.152.152.152h5.582V8.248zm4.994.267h-.017c-.084.333-.166.755-.258 1.078l-.332 1.187h1.236l-.347-1.187c-.1-.332-.2-.746-.282-1.078z"
      />
      <path
        d="M15.152 5.841H6.41a.152.152 0 0 0-.152.152v8.743c0 .084.068.151.152.151h8.742a.15.15 0 0 0 .152-.151V5.993a.152.152 0 0 0-.152-.152zm-3.18 7.319l-.431-1.436H9.94l-.399 1.436h-1.31L9.94 7.57h1.658l1.733 5.59h-1.359zM3.935 9.846H2.72v2.02l1.006.996h2.26v-1.215H3.935z"
      />
    </g>
  </SVGIcon>
  <Button__Label>
    Test
  </Button__Label>
</Button__StyledButton>
`;

exports[`<Button /> should render and match the snapshot 1`] = `
<Button__StyledButton
  aria-expanded={false}
  aria-haspopup="true"
  innerRef={[MockFunction]}
  onClick={[Function]}
  onKeyDown={[Function]}
>
  <SVGIcon
    viewBox="0 0 15.304 15.304"
  >
    <g>
      <path
        d="M5.734 8.248v-.631a5.587 5.587 0 0 1-.992-.863 12.413 12.413 0 0 1-1.836 1.464l-.081.05-.49-.783.08-.051c.01-.007.876-.556 1.751-1.413a8.345 8.345 0 0 1-1.124-2.402l-.023-.09.895-.23.022.091c.002.01.257.954.877 1.926.613-.745.97-1.463 1.061-2.14H2.115V2.171h2.072v-.559h.966v.558h2.072v1.006h-.417c-.097.941-.575 1.916-1.424 2.901.114.128.231.245.35.357v-.473c0-.336.272-.608.608-.608h2.705V.569a.151.151 0 0 0-.152-.152H.152A.152.152 0 0 0 0 .569v8.742c0 .085.068.152.152.152h5.582V8.248zm4.994.267h-.017c-.084.333-.166.755-.258 1.078l-.332 1.187h1.236l-.347-1.187c-.1-.332-.2-.746-.282-1.078z"
      />
      <path
        d="M15.152 5.841H6.41a.152.152 0 0 0-.152.152v8.743c0 .084.068.151.152.151h8.742a.15.15 0 0 0 .152-.151V5.993a.152.152 0 0 0-.152-.152zm-3.18 7.319l-.431-1.436H9.94l-.399 1.436h-1.31L9.94 7.57h1.658l1.733 5.59h-1.359zM3.935 9.846H2.72v2.02l1.006.996h2.26v-1.215H3.935z"
      />
    </g>
  </SVGIcon>
  <Button__Label>
    Label
  </Button__Label>
</Button__StyledButton>
`;
